cpp/include/raft/neighbors/nn_descent_types.hpp

```diff
@@ -105,6 +108,7 @@ struct index : ann::index {
       res_{res},
       metric_{raft::distance::DistanceType::L2Expanded},
       graph_{raft::make_host_matrix<IdxT, int64_t, row_major>(0, 0)},
+      distances_{raft::make_host_matrix<DistData_t, int64_t, row_major>(0, 0)},
       graph_view_{graph_view}
```

Comment:
Modify this constructor to allow the user to provide a pre-allocated distance matrix, just like how the constructor allows for the graph. The signature can be of the form, `std::optional<raft::host_matrix_view<...>>`.

Also, you'll need a function called `allocate_distances` or something so that when `return_distances == true`, you can allocate memory for the distances.

Reply:
I'm not sure if I understood correctly..! Do you mean
- there should be a member function in `index` struct that allocates the `host_matrix` for `distances_` 
- `distances_` should be declared as a `std::optional` type

Reply:
Yes, that is indeed what I am suggesting.

Reply:
Instead of making another function, I used the struct constructor to allocate the `distances_` (and `distance_view_`) depending on `return_distances` flag